File: tests/evolution.test.ts
import assert from "node:assert/strict";
import { test } from "node:test";
import { cp, mkdir, mkdtemp, readFile, rm, symlink, writeFile } from "node:fs/promises";
import { join, resolve } from "node:path";
import { tmpdir } from "node:os";
import { pathToFileURL } from "node:url";
import { execFileSync } from "node:child_process";
import { createClient } from "@connectrpc/connect";
import { createGrpcTransport } from "@connectrpc/connect-node";
import { startBackends } from "../server/backends.js";
import { serveGrpc } from "../server/network.js";
import { createResolvers } from "../server/resolvers.js";
import { AppService as OldAppService } from "../generated/app_pb.js";
import { readFileSync } from "node:fs";

test("editing a child fragment changes the generated contract and actual gRPC response", async t => {
  const dir = await mkdtemp(join(tmpdir(), "fragment-rpc-evolution-"));
  t.after(() => rm(dir, { recursive:true, force:true }));
  await Promise.all([
    cp("app", join(dir, "app"), { recursive:true }),
    cp("proto", join(dir, "proto"), { recursive:true }),
    cp("schema.graphql", join(dir, "schema.graphql")),
    mkdir(join(dir, "server")),
    mkdir(join(dir, "generated")),
    symlink(resolve("node_modules"), join(dir, "node_modules"), "dir"),
    writeFile(join(dir, "package.json"), '{"type":"module"}'),
  ]);
  await Promise.all([
    cp("generated/schema.lock.json", join(dir, "generated/schema.lock.json")),
    cp("server/execute.ts", join(dir, "server/execute.ts")),
    cp("server/plan.ts", join(dir, "server/plan.ts")),
  ]);
  const fragment = join(dir, "app/components/UserCard/UserCard.graphql");
  await writeFile(fragment, (await readFile(fragment, "utf8")).replace("  name\n", "  name\n  email\n"));
  execFileSync(process.execPath, ["--import", "tsx", resolve("compiler/generate.ts")], { cwd:dir, stdio:"pipe" });
  const proto = await readFile(join(dir, "generated/app.proto"), "utf8");
  assert.match(proto, /string email = 5;/);
  assert.match(await readFile(join(dir, "generated/fragments.ts"), "utf8"), /email: string/);

  // Runtime only receives generated artifacts and the small plan executor.
  await Promise.all([
    rm(join(dir, "app"), { recursive:true }),
    rm(join(dir, "schema.graphql")),
  ]);
  const { registerOperations } = await import(pathToFileURL(join(dir, "generated/routes.ts")).href);
  const { AppService } = await import(pathToFileURL(join(dir, "generated/app_pb.ts")).href);
  const backends = await startBackends();
  t.after(() => backends.close());
  const gateway = await serveGrpc(router => registerOperations(router, createResolvers(backends.user.url, backends.organization.url)));
  t.after(() => gateway.close());
  const transport = createGrpcTransport({ baseUrl:gateway.url });
  const client = createClient(AppService as typeof OldAppService, transport);
  const response = await client.getUserPage({ id:"u1" });
  assert.equal(Reflect.get(response.user!, "email"), "aki@example.test");
  const oldClient = createClient(OldAppService, transport);
  const oldResponse = await oldClient.getUserPage({ id:"u1" });
  assert.equal(oldResponse.user?.name, "Aki Tanaka");
  assert.equal(oldResponse.user?.organization?.name, "Northstar Studio");
  assert.equal(Object.hasOwn(oldResponse.user!, "email"), false);
});

test("runtime source graph does not import the GraphQL compiler", () => {
  for (const file of ["server/dev.ts", "server/execute.ts", "server/resolvers.ts", "generated/routes.ts", "generated/app_pb.ts"]) {
    assert.doesNotMatch(readFileSync(file, "utf8"), /from ["']graphql["']|from ["'].*compiler\//);
  }
});

File: internal/demo/evolution_test.go
package demo_test

import (
	"bufio"
	"bytes"
	"context"
	"encoding/json"
	"io/fs"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"syscall"
	"testing"
	"time"

	"connectrpc.com/connect"
	"github.com/azrsh/fragment-colocation-with-grpc/generated/appv1"
	"github.com/azrsh/fragment-colocation-with-grpc/generated/appv1/appv1connect"
	"github.com/azrsh/fragment-colocation-with-grpc/internal/compiler"
	"github.com/azrsh/fragment-colocation-with-grpc/internal/network"
	"google.golang.org/protobuf/proto"
)

func TestFragmentEditChangesStandaloneRPCAndPreservesOldClient(t *testing.T) {
	root, err := filepath.Abs("../..")
	if err != nil {
		t.Fatal(err)
	}
	dir := t.TempDir()
	for _, entry := range []string{"app", "proto", "schema.graphql", "go.mod", "go.sum", "internal", "cmd/server", "generated/schema.lock.json"} {
		err := filepath.WalkDir(filepath.Join(root, entry), func(path string, e fs.DirEntry, err error) error {
			if err != nil {
				return err
			}
			if e.IsDir() {
				return nil
			}
			relative, err := filepath.Rel(root, path)
			if err != nil {
				return err
			}
			data, err := os.ReadFile(path)
			if err != nil {
				return err
			}
			target := filepath.Join(dir, relative)
			if err = os.MkdirAll(filepath.Dir(target), 0755); err != nil {
				return err
			}
			return os.WriteFile(target, data, 0644)
		})
		if err != nil {
			t.Fatal(err)
		}
	}
	if err = os.Symlink(filepath.Join(root, "node_modules"), filepath.Join(dir, "node_modules")); err != nil {
		t.Fatal(err)
	}
	fragment := filepath.Join(dir, "app/components/UserCard/UserCard.graphql")
	data, err := os.ReadFile(fragment)
	if err != nil {
		t.Fatal(err)
	}
	if err = os.WriteFile(fragment, []byte(strings.Replace(string(data), "  name\n", "  name\n  email\n", 1)), 0644); err != nil {
		t.Fatal(err)
	}
	if _, err = compiler.Generate(dir); err != nil {
		t.Fatal(err)
	}
	data, err = os.ReadFile(filepath.Join(dir, "generated/app.proto"))
	if err != nil {
		t.Fatal(err)
	}
	if !bytes.Contains(data, []byte("string email = 5;")) {
		t.Fatalf("missing new field: %s", data)
	}
	data, err = os.ReadFile(filepath.Join(dir, "generated/fragments.ts"))
	if err != nil {
		t.Fatal(err)
	}
	if !bytes.Contains(data, []byte("email: string")) {
		t.Fatal("missing fragment field")
	}
	deps := exec.Command("go", "list", "-deps", "./cmd/server")
	deps.Dir = dir
	data, err = deps.CombinedOutput()
	if err != nil {
		t.Fatalf("dependencies: %v: %s", err, data)
	}
	if bytes.Contains(data, []byte("gqlparser")) || bytes.Contains(data, []byte("/internal/compiler")) {
		t.Fatal("runtime imports GraphQL compiler")
	}
	runtimeDir := t.TempDir()
	binary := filepath.Join(runtimeDir, "fragment-rpc")
	build := exec.Command("go", "build", "-o", binary, "./cmd/server")
	build.Dir = dir
	if data, err = build.CombinedOutput(); err != nil {
		t.Fatalf("build evolved server: %v: %s", err, data)
	}
	// The executable must serve the new contract after all generation inputs are gone.
	if err = os.RemoveAll(dir); err != nil {
		t.Fatal(err)
	}
	cmd := exec.Command(binary)
	cmd.Dir = runtimeDir
	for _, env := range os.Environ() {
		if !strings.HasPrefix(env, "PATH=") && !strings.HasPrefix(env, "PORT=") && !strings.HasPrefix(env, "GRPC_PORT=") {
			cmd.Env = append(cmd.Env, env)
		}
	}
	cmd.Env = append(cmd.Env, "PATH="+runtimeDir, "PORT=0", "GRPC_PORT=0")
	var stderr bytes.Buffer
	cmd.Stderr = &stderr
	stdout, err := cmd.StdoutPipe()
	if err != nil {
		t.Fatal(err)
	}
	if err = cmd.Start(); err != nil {
		t.Fatal(err)
	}
	lines := make(chan string, 8)
	scanDone := make(chan struct{})
	go func() {
		defer close(lines)
		defer close(scanDone)
		scanner := bufio.NewScanner(stdout)
		for scanner.Scan() {
			line := scanner.Text()
			if strings.HasPrefix(line, "Demo:") || strings.HasPrefix(line, "Generated gRPC API:") {
				lines <- line
			}
		}
	}()
	t.Cleanup(func() {
		cmd.Process.Signal(syscall.SIGTERM)
		done := make(chan error, 1)
		go func() { done <- cmd.Wait() }()
		select {
		case err := <-done:
			if err != nil {
				t.Errorf("server exit: %v: %s", err, stderr.String())
			}
		case <-time.After(5 * time.Second):
			cmd.Process.Kill()
			<-done
			t.Error("server did not stop")
		}
		<-scanDone
	})
	webURL, grpcURL := "", ""
	timeout := time.After(10 * time.Second)
	for webURL == "" || grpcURL == "" {
		select {
		case line, ok := <-lines:
			if !ok {
				t.Fatal("server exited before startup")
			}
			if strings.HasPrefix(line, "Demo: ") {
				webURL = strings.TrimPrefix(line, "Demo: ")
			}
			if strings.HasPrefix(line, "Generated gRPC API: ") {
				grpcURL = strings.TrimSuffix(strings.TrimPrefix(line, "Generated gRPC API: "), "/app.v1.AppService")
			}
		case <-timeout:
			t.Fatal("server startup timed out")
		}
	}
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, webURL+"/app.v1.AppService/GetUserPage", strings.NewReader(`{"id":"u1"}`))
	if err != nil {
		t.Fatal(err)
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Connect-Protocol-Version", "1")
	response, err := http.DefaultClient.Do(req)
	if err != nil {
		t.Fatal(err)
	}
	defer response.Body.Close()
	if response.StatusCode != 200 {
		t.Fatal(response.Status)
	}
	var payload struct{ User struct{ Email string } }
	if err = json.NewDecoder(response.Body).Decode(&payload); err != nil {
		t.Fatal(err)
	}
	if payload.User.Email != "aki@example.test" {
		t.Fatalf("evolved response: %+v", payload)
	}
	httpClient := network.NewGRPCClient()
	defer httpClient.CloseIdleConnections()
	oldClient := appv1connect.NewAppServiceClient(httpClient, grpcURL, connect.WithGRPC())
	old, err := oldClient.GetUserPage(ctx, connect.NewRequest(&appv1.GetUserPageRequest{Id: proto.String("u1")}))
	if err != nil {
		t.Fatal(err)
	}
	if old.Msg.GetUser().GetName() != "Aki Tanaka" || old.Msg.GetUser().GetOrganization().GetName() != "Northstar Studio" {
		t.Fatal(old.Msg)
	}
	if old.Msg.User.ProtoReflect().Descriptor().Fields().ByName("email") != nil {
		t.Fatal("old client unexpectedly has new field")
	}
}
